skyportal/handlers/api/photometry.py

```diff
@@ -182,46 +202,170 @@ def post(self):
+                                      f'missing required field {field}.')
+
+            # convert the mags to fluxes
+            # detections
+            detflux = 10**(-0.4 * (df[magdet]['mag'] - PHOT_ZP))
+            detfluxerr = df[magdet]['magerr'] / (2.5 / np.log(10)) * detflux
+
+            # non-detections
+            limmag_flux = 10**(-0.4 * (df[magnull]['limiting_mag'] - PHOT_ZP))
+            ndetfluxerr = limmag_flux / df[magnull]['limiting_mag_nsigma']
+
+            # initialize flux to be none
+            phot_table = Table.from_pandas(df[['mjd', 'magsys', 'filter']])
+
+            phot_table['zp'] = PHOT_ZP
+            phot_table['flux'] = np.nan
+            phot_table['fluxerr'] = np.nan
+            phot_table['flux'][magdet] = detflux
+            phot_table['fluxerr'][magdet] = detfluxerr
+            phot_table['fluxerr'][magnull] = ndetfluxerr
 
+        else:
+            for field in PhotFluxFlexible.required_keys:
+                missing = df[field].isna()
+                if any(missing):
+                    first_offender = np.argwhere(missing)[0, 0]
+                    packet = df.iloc[first_offender].to_dict()
+
+                    for key in packet:
+                        packet[key] = nan_to_none(packet[key])
+
+                    return self.error(f'Error parsing packet "{packet}": '
+                                      f'missing required field {field}.')
+
+            phot_table = Table.from_pandas(df[['mjd', 'magsys', 'filter', 'zp']])
+            phot_table['flux'] = df['flux'].fillna(np.nan)
+            phot_table['fluxerr'] = df['fluxerr'].fillna(np.nan)
+
+        # convert to microjanskies, AB for DB storage as a vectorized operation
+        pdata = PhotometricData(phot_table)
+        standardized = pdata.normalized(zp=PHOT_ZP, zpsys='ab')
+
+        df['standardized_flux'] = standardized.flux
+        df['standardized_fluxerr'] = standardized.fluxerr
+
+        instcache = {}
+        for iid in df['instrument_id'].unique():
+            instrument = Instrument.query.get(int(iid))
+            if not instrument:
+                return self.error(
+                    f'Invalid instrument ID: {iid}')
+            instcache[iid] = instrument
+
+        for oid in df['obj_id'].unique():
+            obj = Obj.query.get(oid)
+            if not obj:
+                return self.error(f'Invalid object ID: {oid}')
+
+        # pre-fetch the photometry PKs. these are not guaranteed to be
+        # gapless (e.g., 1, 2, 3, 4, 5, ...) but they are guaranteed
+        # to be unique in the table and thus can be used to "reserve"
+        # PK slots for uninserted rows
+        pkq = f"SELECT nextval('photometry_id_seq') FROM " \
+              f"generate_series(1, {len(df)})"
+
+        proxy = DBSession().execute(pkq)
+
+        # cache this as list for response
+        ids = [i[0] for i in proxy]
+        df['id'] = ids
+        rows = df.where(pd.notnull(df), None).to_dict('records')
```

Comment:
Do you really need to convert df to a dict? Using `df.loc[whatever].itertuples()` on L:324 should be much faster

Reply:
Need dicts for L:358

Reply:
Sure, but you can iterate `for row in df.where(pd.notnull(df), None).itertuples()` and assemble `phot` using the named tuple `row`, e.g. L:330 would be `flux = row.standardized_flux`. It's usually much faster, but I don't think we'll notice the difference in most cases.